runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingEngineWorkCommitterTest.java

```diff
@@ -0,0 +1,242 @@
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+@RunWith(JUnit4.class)
+public class StreamingEngineWorkCommitterTest {
+
+  @Rule public ErrorCollector errorCollector = new ErrorCollector();
+  private StreamingEngineWorkCommitter workCommitter;
+  private FakeWindmillServer fakeWindmillServer;
+  private Supplier<CloseableStream<CommitWorkStream>> commitWorkStreamFactory;
+
+  private static Work createMockWork(long workToken, Consumer<Work> processWorkFn) {
+    return Work.create(
+        Windmill.WorkItem.newBuilder()
+            .setKey(ByteString.EMPTY)
+            .setWorkToken(workToken)
+            .setShardingKey(workToken)
+            .setCacheToken(workToken)
+            .build(),
+        Instant::now,
+        Collections.emptyList(),
+        processWorkFn);
+  }
+
+  private static ComputationState createComputationState(String computationId) {
+    return new ComputationState(
+        computationId,
+        new MapTask().setSystemName("system").setStageName("stage"),
+        Mockito.mock(BoundedQueueExecutor.class),
+        ImmutableMap.of(),
+        null);
+  }
+
+  private static CompleteCommit asCompleteCommit(Commit commit, Windmill.CommitStatus status) {
+    if (commit.work().isFailed()) {
+      return CompleteCommit.forFailedWork(commit);
+    }
+
+    return CompleteCommit.create(commit, status);
+  }
+
+  @Before
+  public void setUp() throws IOException {
+    fakeWindmillServer =
+        new FakeWindmillServer(
+            errorCollector, ignored -> Optional.of(Mockito.mock(ComputationState.class)));
+    commitWorkStreamFactory =
+        WindmillStreamPool.create(
+                1, Duration.standardMinutes(1), fakeWindmillServer::commitWorkStream)
+            ::getCloseableStream;
+  }
+
+  @After
+  public void cleanUp() {
+    workCommitter.stop();
+  }
+
+  private StreamingEngineWorkCommitter createWorkCommitter(
+      Consumer<CompleteCommit> onCommitComplete) {
+    return StreamingEngineWorkCommitter.create(commitWorkStreamFactory, 1, onCommitComplete);
+  }
+
+  @Test
+  public void testCommit_sendsCommitsToStreamingEngine() {
+    Set<CompleteCommit> completeCommits = new HashSet<>();
+    workCommitter = createWorkCommitter(completeCommits::add);
+    List<Commit> commits = new ArrayList<>();
+    for (int i = 1; i <= 5; i++) {
+      Work work = createMockWork(i, ignored -> {});
+      WorkItemCommitRequest commitRequest =
+          WorkItemCommitRequest.newBuilder()
+              .setKey(work.getWorkItem().getKey())
+              .setShardingKey(work.getWorkItem().getShardingKey())
+              .setWorkToken(work.getWorkItem().getWorkToken())
+              .setCacheToken(work.getWorkItem().getCacheToken())
+              .build();
+      commits.add(Commit.create(commitRequest, createComputationState("computationId-" + i), work));
+    }
+
+    workCommitter.start();
+    commits.forEach(workCommitter::commit);
```

Comment:
use a paralleliszed forEach to cover thread-safety a little?

Reply:
done